### DISCUSSION TITLE
[OPEN BETA] New installation script

### DISCUSSION THREAD

**Discussion by Staubgeborener** [Staubgeborener/klipper-backup#93] [General]
@Tylerjet and I are happy to announce that after a lot of work the **installation script** is going into open beta. Before I go into this, I would like to thank @Tylerjet, who did 99% of the work. :tada:

# :bulb: The idea
The idea behind this script is to automate everything as much as possible, as the manual installation, especially the addition of new features such as cronjobs or systemd services, led to errors for some users.
The script should now guide the user through the most important installation steps and enables us to integrate future features more easily, such as the [official KIAUH support](https://github.com/dw-0/kiauh/issues/435).

The script is currently located in a separate branch called [installer-beta](https://github.com/Staubgeborener/klipper-backup/tree/installer-beta). After the public beta, everything will be merged into the main branch. 

Every user who participates here takes a huge amount of test work off our hands, as there are many scenarios that we simply can't think of.

# :gift_heart: How to participate
**Simply try the installation script to install klipper-backup.**

Use it to install various services such as [klipper-backup-on-boot.service](https://staubgeborener.github.io/klipper-backup/automation/#backup-on-boot) or [klipper-backup-filewatch.service](https://staubgeborener.github.io/klipper-backup/automation/#backup-on-file-changes) or have a [cronjob](https://staubgeborener.github.io/klipper-backup/automation/#cron) created automatically. Have a [moonraker entry](https://staubgeborener.github.io/klipper-backup/updating/#moonraker-update-manager) added automatically or not. And so on. Try different constellations and check whether the script actually executes what you expect. 

If there is a mistake somewhere - tell us!
If everything is going well - tell us!

All we need is feedback.

# :calendar: How long does the beta last?

I hope only a few weeks. The more data and information we get, the better and faster the beta should be over and everything will be merged into the main branch.

# :pencil: How to use
To use the script and thus participate in the public beta, simply use the following commands in the terminal:

```shell
cd ~
cp ./klipper-backup/.env . # create a backup of your .env file in your home directory - just in case!
sudo rm -r ./klipper-backup ./config_backup # we want a clean start so delete the old files
git clone https://github.com/Staubgeborener/klipper-backup
cd ./klipper-backup
git checkout installer-beta
```

Start installation script: `~/klipper-backup/install.sh` (you can still use some of your old .env files: `cat ~/.env`)

**Comment by BenBergman** [comment]
Needed to add klipper-backup to a new printer, so this is perfect timing for me. I'll put my notes here.

1. ~~On the GitHub token page, it would be nice if it gave the info for how to generate that or provided a link to that info.~~
2. ~~When installing the file watch service, I got this error message and it wasn't clear if it self recovered or not~~
    - `/home/biqu/klipper-backup/utils/utils.func: line 189: inotifywait: command not found`
    - Running through the installer again did not give me the same error, but changing my `printer.cfg` isn't pushing new commits to GitHub
    - I have tried doing `sudo apt-get install inotify-tools`
3. ~~Doing `chmod +x *` changed the permissions on the readme, so moonraker update thinks my repo is dirty now.~~
4. ~~On-boot service is not pushing my configs on reboot~~
5. ~~Filewatch commit message template isn't filled in correctly - `printer.cfg modified - %x - %X`~~

Sorry if the constant edits are making this hard to follow live. I figure it is better than me sending a million messages though.

**Reply by Tylerjet** [reply]
By page you just mean the insert token step of the installer right?

**Reply by BenBergman** [reply]
Correct

**Reply by Tylerjet** [reply]
@BenBergman you should be able to ```git pull``` i have added an echo that gives the link to githubs docs on creating tokens

**Reply by Tylerjet** [reply]
for 3.) yeah i don't actually think chmod is needed as all the scripts are already executable I'll edit the post above for that. As for 2.) i should be able to check that in a bit seems like i probably messed up the order of operations on some checks for if inotify is installed

**Reply by Tylerjet** [reply]
also could you post the systemd status of on-boot service and the filewatch service

**Reply by BenBergman** [reply]
On-boot:
``` 
● klipper-backup-on-boot.service - Klipper Backup On-boot Service
     Loaded: loaded (/etc/systemd/system/klipper-backup-on-boot.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Tue 2024-04-16 15:38:02 CDT; 4min 4s ago
    Process: 1103 ExecStart=/usr/bin/env bash -c /usr/bin/env bash $HOME/klipper-backup/script.sh "New Backup on boot - $(date +"%x - %X")" (code=exited, statu>
   Main PID: 1103 (code=exited, status=0/SUCCESS)
        CPU: 2.281s

Apr 16 15:38:00 BTT-CB1 env[1103]: Skipping symbolic link: printer_data/config/timelapse.cfg
Apr 16 15:38:01 BTT-CB1 env[1452]: On branch main
Apr 16 15:38:01 BTT-CB1 env[1452]: nothing to commit, working tree clean
Apr 16 15:38:01 BTT-CB1 env[1461]: remote: Write access to repository not granted.
Apr 16 15:38:01 BTT-CB1 env[1461]: fatal: unable to access 'https://github.com/BenBergman/klipper-config-micron-plus.git/': The requested URL returned error: 4>
Apr 16 15:38:02 BTT-CB1 env[1468]: remote: Write access to repository not granted.
Apr 16 15:38:02 BTT-CB1 env[1468]: fatal: unable to access 'https://github.com/BenBergman/klipper-config-micron-plus.git/': The requested URL returned error: 4>
Apr 16 15:38:02 BTT-CB1 systemd[1]: klipper-backup-on-boot.service: Succeeded.
Apr 16 15:38:02 BTT-CB1 systemd[1]: Finished Klipper Backup On-boot Service.
Apr 16 15:38:02 BTT-CB1 systemd[1]: klipper-backup-on-boot.service: Consumed 2.281s CPU time.
```

Filewatch:
```
● klipper-backup-filewatch.service - Klipper Backup Filewatch Service
     Loaded: loaded (/etc/systemd/system/klipper-backup-filewatch.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Tue 2024-04-16 15:37:56 CDT; 4min 54s ago
    Process: 1100 ExecStart=/usr/bin/env bash -c       watchlist="";      while IFS= read -r path; do          for file in $path; do              if [ ! -h "$f>
' | sort -u | tr '
' ' ');      exclude_pattern=".swp|.tmp|printer-[0-9]*_[0-9]*.cfg|.bak|.bkp";      inotifywait -mrP -e close_write -e move -e delete --exclude "$exclude_patter>
   Main PID: 1100 (code=exited, status=0/SUCCESS)
        CPU: 69ms

Apr 16 15:37:56 BTT-CB1 systemd[1]: Started Klipper Backup Filewatch Service.
Apr 16 15:37:56 BTT-CB1 env[1131]: inotifywait: invalid option -- 'P'
Apr 16 15:37:56 BTT-CB1 systemd[1]: klipper-backup-filewatch.service: Succeeded.
```

**Reply by Tylerjet** [reply]
ahh ok so for on-boot you might need to recreate the token, or if you can edit the token make sure it has access to that specific repo and has pull/push perms.

**Reply by BenBergman** [reply]
I copied the token from my other printer and that token should have access to both repos. I'll try making a new token just for this printer - give it a fresh start.

**Reply by Tylerjet** [reply]
and then for filewatch if you run ```inotifywait -h``` what is the version number reported as?

**Reply by BenBergman** [reply]
Derp - I forgot to hit save when adding permission to the new repo on the old token. New token is working with on-boot now.

Re `inotifywait -h`:

```
inotifywait 3.14                                                                
Wait for a particular event on a file or set of files.
Usage: inotifywait [ options ] file1 [ file2 ] [ file3 ] [ ... ]       
Options:                              
        -h|--help       Show this help text.                                   
        @<file>         Exclude the specified file from being watched.
        --exclude <pattern>                                                     
                        Exclude all events on files matching the      
                        extended regular expression <pattern>.              
        --excludei <pattern>                                                    
                        Like --exclude but case insensitive.              
        -m|--monitor    Keep listening for events forever.  Without
                        this option, inotifywait will exit after one      
                        event is received.
        -d|--daemon     Same as --monitor, except run in the background
                        logging events to a file specified by --outfile.
                        Implies --syslog.
        -r|--recursive  Watch directories recursively.
        --fromfile <file>
                        Read files to watch from <file> or `-' for stdin.
        -o|--outfile <file>
                        Print events to <file> rather than stdout.
        -s|--syslog     Send errors to syslog rather than stderr.
        -q|--quiet      Print less (only print events).
        -qq             Print nothing (not even events).
        --format <fmt>  Print using a specified printf-like format
                        string; read the man page for more details.
        --timefmt <fmt> strftime-compatible format string for use with
                        %T in --format string.
        -c|--csv        Print events in CSV format.
        -t|--timeout <seconds>
                        When listening for a single event, time out after
                        waiting for an event for <seconds> seconds.
                        If <seconds> is 0, inotifywait will never time out.
        -e|--event <event1> [ -e|--event <event2> ... ]
                Listen for specific event(s).  If omitted, all events are 
                listened for.

Exit status:
        0  -  An event you asked to watch for was received.
        1  -  An event you did not ask to watch for was received
              (usually delete_self or unmount), or some error occurred.
        2  -  The --timeout option was given and no events occurred
              in the specified interval of time.

Events:
        access          file or directory contents were read
        modify          file or directory contents were written
        attrib          file or directory attributes changed
        close_write     file or directory closed, after being opened in
                        writable mode
        close_nowrite   file or directory closed, after being opened in
                        read-only mode
        close           file or directory closed, regardless of read/write mode
        open            file or directory opened
        moved_to        file or directory moved to watched directory
        moved_from      file or directory moved from watched directory
        move            file or directory moved to or from watched directory
        create          file or directory created within watched directory
        delete          file or directory deleted within watched directory
        delete_self     file or directory was deleted
        unmount         file system containing file or directory unmounted
```

FYI, I'm running BigTreeTech's OS for CB1.

**Reply by Tylerjet** [reply]
hrmm ok when running the filewatch installer does it prompt and state its installing inotify and building it

**Reply by BenBergman** [reply]
Ran `install.sh` again and it did do a build. Now my inotifywait is 4.23.9.0. I probably installed inotify-tools from the OS repository between my first install and getting you those errors and version numbers.

That said, I'm still getting an error, but it has changed:

```
● klipper-backup-filewatch.service - Klipper Backup Filewatch Service
     Loaded: loaded (/etc/systemd/system/klipper-backup-filewatch.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Tue 2024-04-16 16:05:43 CDT; 26s ago
    Process: 1094 ExecStart=/usr/bin/env bash -c       watchlist="";      while IFS= read -r path; do          for file in $path; do              if [ ! -h "$f>
' | sort -u | tr '
' ' ');      exclude_pattern=".swp|.tmp|printer-[0-9]*_[0-9]*.cfg|.bak|.bkp";      inotifywait -mrP -e close_write -e move -e delete --exclude "$exclude_patter>
   Main PID: 1094 (code=exited, status=0/SUCCESS)
        CPU: 74ms

Apr 16 16:05:43 BTT-CB1 systemd[1]: Started Klipper Backup Filewatch Service.
Apr 16 16:05:43 BTT-CB1 env[1137]: inotifywait: symbol lookup error: inotifywait: undefined symbol: inotifytools_init
Apr 16 16:05:43 BTT-CB1 systemd[1]: klipper-backup-filewatch.service: Succeeded.
```

**Reply by Tylerjet** [reply]
Gotcha cool, Just working out another function so that if the apt/os installed variant is found we uninstall that first before trying to install the latest build. As for the undefined symbol error can you see if libinotifytools-dev is installed seems like that is where that error comes up.

**Reply by BenBergman** [reply]
According to dpkg/apt, libinotifytools0 is installed. libinotifytools0-dev is available. Nothing without the 0 though.

**Reply by Tylerjet** [reply]
ahh ok lets try installing the one with 0-dev.

**Reply by Tylerjet** [reply]
ahh yep should be the one with a zero as i have that installed
![image](https://github.com/Staubgeborener/klipper-backup/assets/12002714/84e3c5dc-f49c-44ba-bd4f-5c467a2a1f2c)

**Reply by BenBergman** [reply]
I've got it installed. Mine is version 3.14-8.1. Did a reboot to make sure it was getting used, but still getting the same systemctl error message.

**Reply by Tylerjet** [reply]
hrmm do file changes still get detected

**Reply by Tylerjet** [reply]
oh actually i need to update a part of the filewatch service anyways it is still looking for the old v1 paths and has not been updated to the v2 version. But still interested to see if the error actually affects file checking.

**Reply by BenBergman** [reply]
It is not triggering, but I think the v2 file path version is the issue. I did not edit the .env manually and the path format looks quite different from the setup on my old printer.

**Reply by Tylerjet** [reply]
Yeah there is a snazzy little v1 convert script i created that does all the changes automagically but i totally goofed on not updating the service lol

**Reply by Tylerjet** [reply]
@BenBergman Try the installer again and update, or git pull then run the installer, and then check the service and see what the status outputs

**Reply by BenBergman** [reply]
Did a pull, fresh install, then edited printer.cfg. Still getting this error and seemingly not triggering on content change.

```
systemctl status klipper-backup-filewatch.service 
● klipper-backup-filewatch.service - Klipper Backup Filewatch Service
     Loaded: loaded (/etc/systemd/system/klipper-backup-filewatch.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Tue 2024-04-16 17:30:09 CDT; 1min 47s ago
   Main PID: 21141 (code=exited, status=0/SUCCESS)
        CPU: 55ms

Apr 16 17:30:09 BTT-CB1 systemd[1]: Started Klipper Backup Filewatch Service.
Apr 16 17:30:09 BTT-CB1 env[21149]: inotifywait: symbol lookup error: inotifywait: undefined symbol: inotifytools_init
Apr 16 17:30:09 BTT-CB1 systemd[1]: klipper-backup-filewatch.service: Succeeded.
```

**Reply by Tylerjet** [reply]
dangit

**Reply by Tylerjet** [reply]
What flavor of linux is the BTT OS running

**Reply by BenBergman** [reply]
It appears to be based on Debian Bullseye

**Reply by Tylerjet** [reply]
Huzahhh i think i may have found the culprit just doing some final tests to be sure what i did actually works

**Reply by Tylerjet** [reply]
@BenBergman ok really odd request can you run ```sudo apt remove libinotifytools0-dev libinotifytools0```
and then try restarting the filewatch service

you should hopefully see something like this and not get an error from inotifywait
<img width="828" alt="image" src="https://github.com/Staubgeborener/klipper-backup/assets/12002714/c1af3e2d-aed0-4c48-ad0f-6c62abe46baf">

**Reply by Staubgeborener** [reply]
> What flavor of linux is the BTT OS running

I guess it's this [debian 11](https://github.com/bigtreetech/CB1/releases).

**Reply by BenBergman** [reply]
Success! I also had to `sudo rm -rf /usr/include/inotifytools/` as apt was failing to uninstall by itself. But after that and a restart of the filewatch service, all seems to be working now!

**Reply by Tylerjet** [reply]
If you could just go through the original message and strike out what we solved just so i don't come back to it later thinking we didn't fix them lol

**Reply by BenBergman** [reply]
Oh, one issue I am seeing withe the filewatch commit - it seems the template didn't fill out correctly. The commit message is `printer.cfg modified - %x - %X` while the on-boot message is `New Backup on boot - 04/16/2024 - 03:10:14 PM`

**Reply by Tylerjet** [reply]
ooop that is probably another Tyler oppsie when i moved it to its own .sh file i forgot to remove the extra escape characters

**Reply by Tylerjet** [reply]
un momento

**Reply by Tylerjet** [reply]
@BenBergman alright pushed an update you can pull and try it now along with added some checks that ensure those libinotifyfiles are apt removed if they exists.

**Reply by BenBergman** [reply]
filewatch commit messages look good now. Thanks for all the rapid help and fixes!

**Comment by Kellwaxx11** [comment]
Totally new to klipper-backup
I tried the beta installer and everything works fine
- installed without any issues
- files backed up , also on file change triggered ok and backed up

am happy to test more things if needed, but for my basic setup/needs its working perfectly.

Thank you very much

**Reply by Staubgeborener** [reply]
Thanks for the feedback. Which device do you use for klipper? And could you also provide the os: `cat /etc/os-release`.

**Reply by Kellwaxx11** [reply]
I am using a very old 2in1 chinese table "Chuwi Hi8 Air" which comes with Windows+Android.
[Chuwi Hi8 Air Specs](https://www.devicespecifications.com/en/model/668549a7)
but I wiped it out and installed Linux Debian on it, so now it is 2in1 tablet Printer Server + KlipperScreen

output of "cat /etc/os-release"
`PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian`

it not high end at all with 2GB RAM, but havent had any issue with it other than slow wifi (and hard time to get as many hardware to work with Linux as possible)

**Reply by Staubgeborener** [reply]
Old Chinese 2in1 tablet - tbh I would not have expected that. But when it works, it works.

**Comment by n4t3y** [comment]
I have installed this beta installation script and I'm having issues backing up:
```
fly@flypilite2:~$ ~/klipper-backup/script.sh
● Checking for installed dependencies Done!

Klipper-backup is up to date

fatal: not a git repository (or any of the parent directories): .git
fatal: not in a git directory
fatal: not in a git directory
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: No remote configured to list refs from.
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
```
Is the output.

Any ideas what the issue might be? I am directing it to this repo: https://github.com/n4t3y/klipper-configuration-backup

**Reply by Tylerjet** [reply]
try recreating the github repo but don't add a license or readme, then on the linux system do the following:
``` 
cd ~
rm -rf config_backup
```

then run script.sh again

**Reply by n4t3y** [reply]
Deleted the repo and did the above, now I am getting the below error:
```fly@flypilite2:~$ ~/klipper-backup/script.sh
● Checking for installed dependencies Done!

Klipper-backup is up to date

fatal: invalid branch name: init.defaultBranch =
fatal: not in a git directory
fatal: not in a git directory
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: No remote configured to list refs from.
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
```

**Reply by Tylerjet** [reply]
when you pulled the beta install script did you run the initial install.sh and configure the .env file?

**Reply by n4t3y** [reply]
I did run the install.sh again but the .env file does not have the header section with the git information anymore, nor does it have any of the back up path information

**Reply by Tylerjet** [reply]
can you copy the .env file from the klipper-backup folder here. make sure to remove your token before pasting.

**Reply by n4t3y** [reply]
Github won't allow me to attach, but this is the content of it:

```

# Backup paths
#  Note: script.sh starts its search in $HOME which is /home/{username}/
# The array accepts folders or files like the following example
# 
#  backupPaths=( \
#  "printer_data/config/*" \
#  "printer_data/config/printer.cfg" \
#  )
#
# Using the above example the script will search for /home/{username}/printer_data/config/* and /home/{username}/printer_data/config/printer.cfg
# When backing up a folder you should always have /* at the end of the path so that files insde the folder are properly searched

backupPaths=( \ 
)

# Array of strings in .gitignore pattern git format https://git-scm.com/docs/gitignore#_pattern_format for files that should not be uploaded to the remote repo
# New additions must be enclosed in double quotes and should follow the pattern format as noted in the above link

exclude=( \ 
)
```

**Reply by n4t3y** [reply]
I know I could copy the information to manually but I guess this does not test this install beta

**Reply by Tylerjet** [reply]
hrmmm interesting

**Reply by n4t3y** [reply]
When I ran the install initially on the setup, it created it correctly, but ran into the issue I posted intially, but when I ran the command you mentioned above, this is what was created instead, even after running the install.sh again

**Reply by Tylerjet** [reply]
could you try deleting the current .env file then do ``` cp .env.bkp .env ``` and then try running script.sh again

**Reply by n4t3y** [reply]
```fly@flypilite2:~$ cp.env.bkp .env
cp.env.bkp: command not found
fly@flypilite2:~$ cp .env.bkp .env
cp: cannot stat '.env.bkp': No such file or directory
fly@flypilite2:~$ cp .env.example .env
cp: cannot stat '.env.example': No such file or directory
```

**Reply by Tylerjet** [reply]
oh did you not previously use the non beta version?

**Reply by Tylerjet** [reply]
oh wait

**Reply by Tylerjet** [reply]
you need to do the command in klipper-backup folder

**Reply by n4t3y** [reply]
No - this is a new install and I couldn't get the non-beta version to work so I found this
It kept saying that install.sh did not exist

**Reply by Tylerjet** [reply]
gotcha in that case cd into klipper-backup and run the ```cp .env.example .env```

**Reply by Tylerjet** [reply]
and then try running install.sh again and do the configuration and then copy paste the .env file like before minus the token just so we can see how it looks before running script.sh

**Reply by n4t3y** [reply]
```github_token=*************************
github_username=n4t3y
github_repository=klipper-configuration-backup
branch_name="main"
commit_username="n4t3"
commit_email="fly@flypilite2-mlkBBgc"

# Backup paths
#  Note: script.sh starts its search in $HOME which is /home/{username}/
# The array accepts folders or files like the following example
# 
#  backupPaths=( \
#  "printer_data/config/*" \
#  "printer_data/config/printer.cfg" \
#  )
#
# Using the above example the script will search for `/home/{username}/printer_data/config/*` and `/home/{username}/printer_data/config/printer.cfg`
# When backing up a folder you should always have `/*` at the end of the path so that files insde the folder are properly searched 

backupPaths=( \
"printer_data/config/*" \
)

# Array of strings in .gitignore pattern git format https://git-scm.com/docs/gitignore#_pattern_format for files that should not be uploaded to the remote repo
# New additions must be enclosed in double quotes and should follow the pattern format as noted in the above link
exclude=( \
"*.swp" \
"*.tmp" \
"printer-[0-9]*_[0-9]*.cfg" \
"*.bak" \
"*.bkp" \
"*.csv" \
"*.zip" \
)```

**Reply by n4t3y** [reply]
I ran the script.sh again and I'm getting the same error as initially
```
fly@flypilite2:~$ ~/klipper-backup/script.sh
● Checking for installed dependencies Done!

Klipper-backup is up to date

fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
Branch: main in .env does not match the currently checked out branch of: .
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not in a git directory
fatal: not in a git directory
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
Skipping symbolic link: printer_data/config/fluidd.cfg
Skipping symbolic link: printer_data/config/mainsail.cfg
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: No remote configured to list refs from.
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
```

**Reply by Tylerjet** [reply]
ok so now try doing 
```
cd ~
rm -rf config_backup
```
then cd back into klipper-backup, Double check that .env still looks correct, and run script.sh

**Reply by n4t3y** [reply]
Awesome it worked!

**Reply by n4t3y** [reply]
Was this a user error or an install error?

**Reply by n4t3y** [reply]
Thanks for your help by the way

**Reply by Tylerjet** [reply]
I'm not sure lol, was this a fresh linux install?
if so would love to see the history of commands ```history``` in the shell **make sure to scrub personal info** and see if maybe anything jumps out like script.sh accidently getting ran before install.sh which may have messed with order of operations.

**Reply by n4t3y** [reply]
I wiped my SD card and reinstalled the image. It's not a RPI so I had to use a custom image (Mellow Fly Pi Lite 2).

I ran the klipper-backup per the docs (non-beta) but it kept saying that install.sh was not found, so then I found this beta install and went through the process, ending up with this error I initially posted about.

I'm happy to provide any information you need to help track down any faults (if any)

**Reply by n4t3y** [reply]
```
fly@flypilite2:~$ history
    1  exit 0
    2  kiauh/kiauh.sh
    3  sudo apt update && sudo apt install git[B
    4  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh
    5  cd ~ && ./klipper-backup/install.sh
    6  /klipper-backup/install.sh
    7  cd ~
    8  ls
    9  cd klipper-backup
   10  ls
   11  script.sh
   12  cd ~
   13  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh
   14  sudo rm klipper-backup
   15  sudo rmdir klipper-backup
   16  sudo rm -r klipper-backup
   17  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh
   18  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh
   19  sudo rm -r klipper-backup
   20  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh
   21  ./klipper-backup/install.sh
   22  sudo reboot
   23  sudo rm -r klipper-backup
   24  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup/tree/installer-beta && chmod +x ./klipper-backup/script.sh
   25  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup/tree/installer-beta/klipper-backup.git && chmod +x ./klipper-backup/script.sh
   26  cd ~ && git clone -b installer-beta https://github.com/Staubgeborener/klipper-backup
   27  chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh
   28  ./klipper-backup/script.sh
   29  ./klipper-backup/install.sh
   30  ~/klipper-backup/script.sh
   31  cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh
   32  cd ~ && chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh
   33  ./klipper-backup/install.sh
   34  ~/klipper-backup/script.sh
   35  cd ~
   36* cp ./klipper-backup/.env .
   37  sudo rm -r ./klipper-backup ./config_backup
   38  git clone https://github.com/Staubgeborener/klipper-backup
   39  cd ./klipper-backup
   40  git checkout installer-beta
   41  install.sh
   42  cd~
   43  cd ~
   44  ~/klipper-backup/install.sh
   45  ~/klipper-backup/script.sh
   46  cd ~
   47  rm -r config_backup
   48  ~/klipper-backup/script.sh
   49  rm -r config_backup
   50  ~/klipper-backup/install.sh
   51  ~/klipper-backup/script.sh
   52  cp.env.bkp .env
   53  cp .env.bkp .env
   54  cp .env.example .env
   55  cd klipper-backup
   56  cp .env.example .env
   57  script.sh
   58  cd ~
   59  ~/klipper-backup/script.sh
   60  ~/klipper-backup/install.sh
   61  ~/klipper-backup/script.sh
   62  cd ~
   63  rm -rf config_backup
   64  cd klipper-backup
   65  ls
   66  cd ~
   67  ~/klipper-backup/script.sh
   68  history
```

**Reply by Tylerjet** [reply]
hrmm so to me it seems like it could have been a combination of doing different steps between each of the branches that probably it didn't like. I think Staubgeborener updated some of the docs preparing to merge this to main so that may have messed up the normal steps that would have worked hehe.

**Reply by n4t3y** [reply]
Haha, oh well, at least it's working now. I did have it working (non-beta) before I wiped the Sd card, but I had issues reinstalling everything so had to wipe and start again. Originally I installed it on 23rd April without any issues, so not sure if there's been recent changes which affected my install this time around

**Reply by Staubgeborener** [reply]
> I think Staubgeborener updated some of the docs preparing to merge this to main so that may have messed up the normal steps that would have worked hehe.

Yeah, that issue is my fault.

In  @n4t3y history you can see at line 4
`cd ~ && git clone https://github.com/Staubgeborener/klipper-backup.git && chmod +x ./klipper-backup/script.sh ./klipper-backup/install.sh`

which is the latest edit in the docs which i did yesterday. There is no `git checkout` command in order to get `install.sh` (and any other installation file).

I wanted to merge the beta branch into the main branch asap and was probably a bit too hasty. Hooopsie.

**Comment by Staubgeborener** [comment]
Thanks to everyone, [v1.3.0](https://github.com/Staubgeborener/klipper-backup/releases/tag/v1.3.0) is now released.

**Comment by Staubgeborener** [comment]
@BenBergman @Kellwaxx11 @n4t3y We started a new [beta](https://github.com/Staubgeborener/Klipper-Backup/discussions/143) for the restore function.
